### PR TITLE
helm: default to cronJob as tls.auto.method

### PR DESCRIPTION
**Pull request by kaworu**
For both Hubble and Clustermesh. Deprecate the previous default `helm` method on the way.

See first task of https://github.com/cilium/cilium/issues/25345.

### CODE REVIEW

**Review by rolinh — APPROVED** [review]
LGTM, thanks for tackling this!

**Review by gandro — APPROVED** [review]
I'm fine with the change as is, but if we make `cronJob` the default, we likely also need to figure out how `helm uninstall` is supposed to work in `cronJob` mode _before_ the feature freeze in a couple of weeks: https://github.com/cilium/cilium-cli/issues/1648

**Comment by giorio94** [comment]
I haven't yet looked through this PR, but FYI https://github.com/cilium/cilium/pull/25261 should be merged soon, and it will remove the latest occurrences of `hubble.tls.ca`.

File: install/kubernetes/cilium/values.yaml.tmpl [hunk 1]
```diff
@@ -994,13 +994,15 @@ hubble:
       # false, the certs for Hubble server need to be provided by setting
       # appropriate values below.
       enabled: true
-      # -- Set the method to auto-generate certificates. Supported values:
+      # -- Sets the method to auto-generate certificates. Supported values:
+      # - cronJob:      This method uses a Kubernetes CronJob the generate
+      #                 any certificates not provided by the user at
+      #                 installation time.
+      # - certmanager:  This method use cert-manager to generate & rotate
```
Comment (giorio94) [inline install/kubernetes/cilium/values.yaml.tmpl]:
```suggestion
      # - certmanager:  This method uses cert-manager to generate & rotate
```

Nit

File: install/kubernetes/cilium/values.yaml.tmpl [hunk 2]
```diff
@@ -2746,13 +2748,15 @@ clustermesh:
         # enable mTLS between clustermesh-apiserver and external workload instances.
         # If set to false, the certs to be provided by setting appropriate values below.
         enabled: true
-        # Sets the method to auto-generate certificates. Supported values:
+        # -- Sets the method to auto-generate certificates. Supported values:
+        # - cronJob:      This method uses a Kubernetes CronJob the generate
+        #                 any certificates not provided by the user at
+        #                 installation time.
+        # - certmanager:  This method use cert-manager to generate & rotate
```
Comment (giorio94) [inline install/kubernetes/cilium/values.yaml.tmpl]:
```suggestion
        # - certmanager:  This method uses cert-manager to generate & rotate
```

Nit

File: install/kubernetes/cilium/values.yaml.tmpl [hunk 3]
```diff
@@ -2771,7 +2771,7 @@ clustermesh:
         # Defaults to none. Commented syntax gives midnight of the first day of every
         # fourth month. For syntax, see
         # https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/#schedule-syntax
-        # schedule: "0 0 1 */4 *"
+        schedule: "0 0 1 */4 *"
```
Comment (giorio94) [inline install/kubernetes/cilium/values.yaml.tmpl:2766]: Nit. The comment above also need to be updated to reflect this change.

**Review by giorio94 — APPROVED** [review]
Thanks! I just left a couple of nits inline. I guess that you should be able to drop the first commit once https://github.com/cilium/cilium/pull/25261 gets merged.

File: install/kubernetes/cilium/values.yaml.tmpl [hunk 4]
```diff
@@ -994,13 +994,15 @@ hubble:
       # false, the certs for Hubble server need to be provided by setting
       # appropriate values below.
       enabled: true
-      # -- Set the method to auto-generate certificates. Supported values:
+      # -- Sets the method to auto-generate certificates. Supported values:
+      # - cronJob:      This method uses a Kubernetes CronJob the generate
```
Comment (zacharysarah) [inline install/kubernetes/cilium/values.yaml.tmpl]: spelling nit
```suggestion
      # - cronJob:      This method uses a Kubernetes CronJob to generate
```

File: install/kubernetes/cilium/values.yaml.tmpl [hunk 5]
```diff
@@ -2746,13 +2748,15 @@ clustermesh:
         # enable mTLS between clustermesh-apiserver and external workload instances.
         # If set to false, the certs to be provided by setting appropriate values below.
         enabled: true
-        # Sets the method to auto-generate certificates. Supported values:
+        # -- Sets the method to auto-generate certificates. Supported values:
+        # - cronJob:      This method uses a Kubernetes CronJob the generate
```
Comment (zacharysarah) [inline install/kubernetes/cilium/values.yaml.tmpl]: Spelling nit
```suggestion
        # - cronJob:      This method uses a Kubernetes CronJob to generate
```

**Review by zacharysarah — APPROVED** [review]
A couple of nits, otherwise LGTM. Approving with the understanding that changes must be made prior to merge.

**Comment by michi-covalent** [comment]
is this test failure related to this pull request? 👀 https://github.com/cilium/cilium/actions/runs/5057980055/jobs/9077443138?pr=25620

**Review by michi-covalent — APPROVED** [review]
let's harmonize 🧘

**Comment by kaworu** [comment]
@gandro @giorio94 thanks! Addressed all the nits/typos and added a couple more cleanup commits, please take another look.

> I'm fine with the change as is, but if we make `cronJob` the default, we likely also need to figure out how `helm uninstall` is supposed to work in `cronJob` mode _before_ the feature freeze in a couple of weeks: [cilium/cilium-cli#1648](https://github.com/cilium/cilium-cli/issues/1648)

> is this test failure related to this pull request? eyes https://github.com/cilium/cilium/actions/runs/5057980055/jobs/9077443138?pr=25620

@gandro @michi-covalent yes so generally speaking my understanding is that the cilium-cli assumes the helm method (see [here](https://github.com/cilium/cilium-cli/blob/3237c9b041d475d62c9f38470c32342486e33066/hubble/relay.go#L230-L245)) and consequently is not compatible with certgen. The best way forward might be to "teach" certgen to the cilium-cli so that we can drop the helm method eventually, but as @gandro pointed out the problem of generated resource ownership described at https://github.com/cilium/cilium-cli/issues/1648 needs to be tackled (both in the context of the cilium-cli and helm if I understand correctly). Thoughts?

**Comment by kaworu** [comment]
Converted to draft until the cilium-cli compatibility issue is sorted out.

File: install/kubernetes/cilium/values.yaml.tmpl [hunk 6]
```diff
@@ -994,12 +994,14 @@ hubble:
       # false, the certs for Hubble server need to be provided by setting
       # appropriate values below.
       enabled: true
-      # -- Set the method to auto-generate certificates. Supported values:
+      # -- Sets the method to auto-generate certificates. Supported values:
+      # - cronJob:      This method uses a Kubernetes CronJob to generate any
```
Comment (joamaki) [inline install/kubernetes/cilium/values.yaml.tmpl:1042]: nit: "cronjob" would match capitalization of "certmanager" and reads better

**Review by joamaki — APPROVED** [review]

File: install/kubernetes/cilium/values.yaml.tmpl (hunk 6 above)
Comment (kaworu) [reply install/kubernetes/cilium/values.yaml.tmpl:1042]: make sense, we could update the value here and do a case-insensitive check in the template for backward compatibility.

**Review by viktor-kurchenko — APPROVED** [review]
LGTM.

**Comment by asauber** [comment]
When using the CLI in Helm mode (soon to be the default mode), Certgen is now used by default when Clustermesh is enabled.

https://github.com/cilium/cilium-cli/blob/main/clustermesh/clustermesh.go#L1893-L1902

File: install/kubernetes/cilium/values.yaml.tmpl (hunk 6 above)
Comment (kaworu) [reply install/kubernetes/cilium/values.yaml.tmpl:1042]: Thinking back about this we have `.Values.certgen.annotations.cronJob`, so we'll keep cronJob for now for consistency with this annotation Helm value.

**Review by gandro — APPROVED** [review]

File: install/kubernetes/cilium/values.yaml.tmpl [hunk 7]
```diff
@@ -1082,12 +1082,14 @@ hubble:
...
+      #                 certificates not provided by the user at
+      #                 installation time.
+      # - certmanager:  This method uses cert-manager to generate & rotate
+      #                 certificates.
       # - helm:         This method uses Helm to generate all certificates.
-      # - cronJob:      This method uses a Kubernetes CronJob the generate any
-      #                 certificates not provided by the user at installation
-      #                 time.
-      # - certmanager:  This method use cert-manager to generate & rotate certificates.
+      #                 Deprecated and will be removed in Cilium v1.15.
```
Comment (giorio94) [inline install/kubernetes/cilium/values.yaml.tmpl]: I guess this should be 1.16 now. The same below, and in the commit message.

**Review by giorio94 — CHANGES_REQUESTED** [review]
Overall /lgtm. Just one comment about the version in which the deprecated method will be removed.

File: install/kubernetes/cilium/values.yaml.tmpl (hunk 7 above)
Comment (kaworu) [reply install/kubernetes/cilium/values.yaml.tmpl]: Good catch thank you!

**Review by giorio94 — APPROVED** [review]

**Comment by gandro** [comment]
Testfailiure in ci-gingko looks legitimate:

```
  Tests upgrade and downgrade from a Cilium stable image to master [It]
  /home/runner/work/cilium/cilium/test/ginkgo-ext/scopes.go:515

  Cilium "1.14" was not able to be deployed
  Expected command: helm rollback cilium 1 --namespace=kube-system 
  To succeed, but it failed:
  Exitcode: 1 
  Err: exit status 1
  Stdout:
   	 
  Stderr:
   	 Error: no Secret with the name "cilium-ca" found
```

**Comment by michi-covalent** [comment]
@kaworu are you trying to get this in before v1.15 feature freeze? 👀 it's 11/30.

**Comment by kaworu** [comment]
> @kaworu are you trying to get this in before v1.15 feature freeze? 👀 it's 11/30.

As @gandro mentionned the test failures are legit, and I don't think I'll have cycles to look into them before Nov 30th sadly.

**Comment by michi-covalent** [comment]
very sad indeed but i understand 🚀🙏

**Comment by joestringer** [comment]
@kaworu @asauber is there a plan to move this PR forward? Otherwise if the PR is linked from the parent issue https://github.com/cilium/cilium/issues/25345 then we can always find the PR again, so we may as well close this.

**Pull request closed without merging**